src.related/mapgen/src/compress/compress.1
.nr LL 5.5i
.ad b
.hy 1
.TH COMPRESS 1 Utility (7/1/88)
.SH NAME
compress, uncompress, zcat \(em compress and expand data
.SH SYNOPSIS
.B compress
[
.B \-f
] [
.B \-v
] [
.B \-c
] [
.B \-V
] [
.B \-b
.I bits
] [
.I "name \&..."
]
.br
.B uncompress
[
.B \-f
] [
.B \-v
] [
.B \-c
] [
.B \-V
] [
.I "name \&..."
]
.br
.B zcat
[
.B \-V
] [
.I "name \&..."
]
.SH DESCRIPTION
.I Compress
reduces the size of the named files using adaptive Lempel-Ziv coding.
Whenever possible,
each file is replaced by one with the extension
.B "\&.Z,"
while keeping the same ownership modes, access and modification times.
If no files are specified, the standard input is compressed to the
standard output.
Compressed files can be restored to their original form using
.I uncompress
or
.I zcat.
.PP
The
.B \-f
option will force compression of
.I name.
This is useful for compressing an entire directory,
even if some of the files do not actually shrink.
If
.B \-f
is not given and
.I compress
is run in the foreground,
the user is prompted as to whether an existing file should be overwritten.
.PP
The
.B \-c
option makes
.I compress/uncompress
write to the standard output; no files are changed.
The nondestructive behavior of
.I zcat
is identical to that of
.I uncompress
.B \-c.
.PP
.I Compress
uses the modified Lempel-Ziv algorithm popularized in
"A Technique for High Performance Data Compression",
Terry A. Welch,
.I "IEEE Computer,"
vol. 17, no. 6 (June 1984), pp. 8-19.
Common substrings in the file are first replaced by 9-bit codes 257 and up.
When code 512 is reached, the algorithm switches to 10-bit codes and
continues to use more bits until the
limit specified by the
.B \-b
flag is reached (default 16).
.I Bits
must be between 9 and 16.  The default can be changed in the source to allow
.I compress
to be run on a smaller machine.
.PP
After the
.I bits
limit is attained,
.I compress
periodically checks the compression ratio.  If it is increasing,
.I compress
continues to use the existing code dictionary.  However,
if the compression ratio decreases,
.I compress
discards the table of substrings and rebuilds it from scratch.  This allows
the algorithm to adapt to the next "block" of the file.
.PP
Note that the
.B \-b
flag is omitted for
.I uncompress,
since the 
.I bits
parameter specified during compression
is encoded within the output, along with
a magic number to ensure that neither decompression of random data nor
recompression of compressed data is attempted. 
.PP
.ne 8
The amount of compression obtained depends on the size of the
input, the number of
.I bits
per code, and the distribution of common substrings.
Typically, text such as source code or English
is reduced by 50\-60%.
Compression is generally much better than that achieved by
Huffman coding (as used in
.IR pack ),
or adaptive Huffman coding
.RI ( compact ),
and takes less time to compute.
.PP
Under the
.B \-v
option,
a message is printed yielding the percentage of
reduction for each file compressed.
.PP
If the
.B \-V
option is specified, the current version and compile options are printed on
.I stderr.
.PP
Exit status is normally 0;
if the last file is larger after (attempted) compression, the status is 2;
if an error occurs, exit status is 1.
.SH "SEE ALSO"
pack(1), compact(1)
.SH "DIAGNOSTICS"
Usage: compress [\-dfvcV] [\-b maxbits] [file ...]
.in +8
Invalid options were specified on the command line.
.in -8
Missing maxbits
.in +8
Maxbits must follow
.BR \-b \.
.in -8
.IR file :
not in compressed format
.in +8
The file specified to
.I uncompress
has not been compressed.
.in -8
.IR file :
compressed with 
.I xx
bits, can only handle 
.I yy
bits
.in +8
.I File
was compressed by a program that could deal with
more 
.I bits
than the compress code on this machine.
Recompress the file with smaller
.IR bits \.
.in -8
.IR file :
already has .Z suffix -- no change
.in +8
The file is assumed to be already compressed.
Rename the file and try again.
.in -8
.IR file :
filename too long to tack on .Z
.in +8
The file cannot be compressed because its name is longer than
12 characters.
Rename and try again.
This message does not occur on BSD systems.
.in -8
.I file
already exists; do you wish to overwrite (y or n)?
.in +8
Respond "y" if you want the output file to be replaced; "n" if not.
.in -8
uncompress: corrupt input
.in +8
A SIGSEGV violation was detected which usually means that the input file has
been corrupted.
.in -8
Compression: 
.I "xx.xx%"
.in +8
Percentage of the input saved by compression.
(Relevant only for
.BR \-v \.)
.in -8
\(em not a regular file: unchanged
.in +8
When the input file is not a regular file,
(e.g. a directory), it is
left unaltered.
.in -8
\(em has 
.I xx 
other links: unchanged
.in +8
The input file has links; it is left unchanged.  See
.IR ln "(1)"
for more information.
.in -8
\(em file unchanged
.in +8
No savings is achieved by
compression.  The input remains virgin.
.in -8
.SH "BUGS"
Although compressed files are compatible between machines with large memory,
.BR \-b \12
should be used for file transfer to architectures with 
a small process data space (64KB or less, as exhibited by the DEC PDP
series, the Intel 80286, etc.)


CHANGES
- GISLIB 
  - G_warning(), G_fatal_error() - send mail only if eviroment variable
                                   GRASS_ERROR_MAIL is set
  - G_debug() - new, print debugging info on selected level
  - G_plot_icon() - new, plot selected mark
- DBMI 
  - dbmscap file is not used, drivers are searched in dist.$ARCH/driver/db
  - new dbf driver
- vector library completely rewritten (multilayer, multiformat, 3D)
- new make system


src/compress/compress.c
/****************************************************************************
 * compress_cell converts straight grid_cell files into compressed grid_cell
 * files.  Compressed files have the following format:
 *
 *  - Array of addresses pointing to the internal start of each row
 *    First byte of each row is the nuber of bytes per cell for that row
 *    Remainder of the row is a series of byte groups that describe the data:
 *        First byte: number of cells that contain the category given by second
 *        Next byte(s): category number. The number of bytes is determined
 *                      by the number of bytes in a cell 
 *
 * The normal G_open_cell(), and G_put_map_row() do the compression
 * This program must only check that the file is not a reclass file and
 * is not already compressed.
 *
 * The only trick is to preserve the support files
 *
 *****************************************************************************/

#include "gis.h"


long newsize, oldsize;
long lseek();

#ifdef DECOMPRESS
#define PROMPT "Enter map to be decompressed"
#define UCNAME "DECOMPRESS"
#define LCNAME "decompress"
#else
#define PROMPT "Enter map to be compressed"
#define UCNAME "COMPRESS"
#define LCNAME "compress"
#endif

main (argc, argv) char *argv[];
{
    int stat ;
    char name[30];

    G_gisinit (argv[0]);

    stat = 0;
    if (argc < 2)
    {
	while (G_ask_cell_in_mapset (PROMPT, name))
	    process (name);
    }
    else
    {
	while (--argc)
	{
	    strcpy (name, *++argv);
	    if (process (name))
		stat = 1;
	}
    }

    exit (stat);
}


process (name)
    char *name;
{
    struct Colors colr;
    struct History hist;
    struct Categories cats;
    int colr_ok;
    int hist_ok;
    int cats_ok;
    long diff;

    printf ("\n%s [%s]\n\n", UCNAME, name);
    if (G_find_cell (name, G_mapset()) == NULL)
    {
	printf ("\7WARNING: [%s] not found\n", name);
	sleep(3);
	return 1;
    }

    G_suppress_warnings(1);
    colr_ok = G_read_colors (name, G_mapset(), &colr) > 0;
    hist_ok = G_read_history (name, G_mapset(), &hist) >= 0;
    cats_ok = G_read_cats (name, G_mapset(), &cats) >= 0;
    G_suppress_warnings(0);

    if (doit(name)) return 1;

    if (colr_ok)
    {
	G_write_colors (name, G_mapset(), &colr);
	G_free_colors (&colr);
    }
    if (hist_ok)
	G_write_history (name, &hist);
    if (cats_ok)
    {
	cats.num = G_number_of_cats (name, G_mapset());
	G_write_cats (name, &cats);
	G_free_cats (&cats);
    }
    diff = newsize - oldsize;
    printf("DONE: %sed file is ", LCNAME);
    if (diff < 0)
    {
	diff = -diff;
	printf("%ld byte%s smaller\n", diff, diff==1?"":"s");
    }
    else if (diff > 0)
    {
	printf("%ld byte%s bigger\n", diff, diff==1?"":"s");
    }
    else
    {
	printf ("same size\n");
    }
    return 0;
}

doit (name)
    char *name;
{
    char rname[30], rmapset[30];
    struct Cell_head cellhd ;
    int new, old, nrows, row;
    CELL *cell;


    if (G_get_cellhd (name, G_mapset(), &cellhd) < 0)
    {
	char msg[100];
	sprintf (msg,"%s: Problem reading cell header for [%s]", UCNAME, name);
	G_warning(msg);
	return 1;
    }
    if (G_is_reclass (name, G_mapset(), rname, rmapset) > 0)
    {
	printf ("[%s] is a reclass file - can't %s\n", name, LCNAME);
	return 1;
    }


/* check if already compressed/decompressed */
#ifdef DECOMPRESS
    if (cellhd.compressed == 0)
#else
    if (cellhd.compressed > 0)
#endif
    {
	printf ("[%s] already %sed\n", name, LCNAME);
	return 1;
    }
    G_set_window (&cellhd);
    old = G_open_cell_old (name, G_mapset());
    if (old < 0)
	return 1;
#ifdef DECOMPRESS
    G_set_cell_format (cellhd.format);
    new = G_open_cell_new_uncompressed (name);
#else
    new = G_open_cell_new (name);
#endif
    if (new < 0)
	return 1;
    nrows = G_window_rows();
    cell = G_allocate_cell_buf();
    oldsize = lseek (old, 0L, 2);

    for (row = 0; row < nrows; row++)
    {
	if (G_get_map_row_nomask (old, cell, row) < 0)
	    break;
	if (G_put_map_row (new, cell) < 0)
	    break;
    }
    free (cell);
    G_close_cell (old);
    if (row < nrows)
    {
	G_unopen_cell (new);
	return 1;
    }
    G_close_cell (new);
    newsize = 0;
    old = G_open_cell_old (name, G_mapset());
    newsize = lseek (old, 0L, 2);
    G_close_cell (old);
    return 0;
}
